Add Manage/Privacy-Delete/Download functionality

HaoK opened the pull request:
Initial changes to identity default UI for Manage/Privacy/DownloadPersonalData and DeletePersonalData

Still need final tweaks to UI/text which we'll plan to get from @DamianEdwards review in person.

@blowdart @danroth27 @javiercn @ajcvickers @mkArtakMSFT

blowdart commented:
There won't be time for in person with @DamianEdwards is off having fun in London, so let's use @danroth27 and I as his proxy.

HaoK commented:
@danroth27 who do you want to review this PR from the perspective of proper RazorPages usage/patterns etc?  I mostly just copied what the existing pages did, I only intentionally deviated slightly using the NavPages constant instead of hardcoding the value in the ActivePages setting.

DamianEdwards commented:
Get Ryan and Pranav to look at it pls

pranavkm commented on src/UI/Areas/Identity/Pages/Account/Manage/DeletePersonalData.cshtml:
```diff
@@ -0,0 +1,28 @@
+</div>
+
+@section Scripts {
+    @await Html.PartialAsync("_ValidationScriptsPartial")
```
`<partial name="_ValidationScriptsPartial" />`

pranavkm replied:
We generally prefer tag helpers to HtmlHelpers. It's cleaner, more idiomatic code.

pranavkm commented on src/UI/Areas/Identity/Pages/Account/Manage/DeletePersonalData.cshtml.cs:
```diff
@@ -0,0 +1,61 @@
+            _logger.LogInformation("User with ID '{UserId}' deleted themselves.", _userManager.GetUserId(User));
+
+            // REVIEW: should this redirect to somewhere that says user was deleted?
+            return RedirectToPage("./");
```
This should redirect to the website root `return Redirect("~/");` or a successfully deleted page. Redirecting to the page `./` is kinda weird.

HaoK replied:
So most things in manage are doing some form of RedirectToPage...  is one better than the other?

pranavkm replied:
RedirectToPage expects to resolve a page. Since there isn't a page named `./`, it ends up resolving the empty string which ends up in the app root. But that's just a weird way of going about it.

HaoK replied:
Yeah so after they get deleted, manage will no longer be valid, so I really was just trying to redirect to home or whatever.  Unless we think its worth adding a dedicated page for a Deleted confirmation

pranavkm commented on src/UI/Areas/Identity/Pages/Account/Manage/DownloadPersonalData.cshtml:
```diff
@@ -0,0 +1,12 @@
+<h4>@ViewData["Title"]</h4>
+
+@section Scripts {
+    @await Html.PartialAsync("_ValidationScriptsPartial")
```
`<partial name="_ValidationScriptsPartial" />`

pranavkm commented on src/UI/Areas/Identity/Pages/Account/Manage/DownloadPersonalData.cshtml.cs:
```diff
@@ -0,0 +1,52 @@
+
+        public DownloadPersonalDataModel(
+            UserManager<IdentityUser> userManager,
+            ILogger<PrivacyModel> logger)
```
`ILogger<DownloadPersonalDataModel>`?

pranavkm commented on src/UI/Areas/Identity/Pages/Account/Manage/DownloadPersonalData.cshtml.cs:
```diff
@@ -0,0 +1,52 @@
+            _logger.LogInformation("User with ID '{UserId}' asked for their personal data.", _userManager.GetUserId(User));
+
+            // REVIEW: do we want to set a force download header?
+            return new FileContentResult(Encoding.ASCII.GetBytes(JsonConvert.SerializeObject(user)), "text/json");
```
You likely need Encoding.UTF8. That said, this kinda feels icky, but there really doesn't seem like a more definitive way to do this.

pranavkm commented on src/UI/Areas/Identity/Pages/Account/Manage/DeletePersonalData.cshtml.cs:
```diff
@@ -0,0 +1,61 @@
+        public DeletePersonalDataModel(
+            UserManager<IdentityUser> userManager,
+            SignInManager<IdentityUser> signInManager,
+            ILogger<PrivacyModel> logger)
```
`ILogger<DeletePersonalDataModel>`

pranavkm commented on src/UI/Areas/Identity/Pages/Account/Manage/DeletePersonalData.cshtml.cs:
```diff
@@ -0,0 +1,61 @@
+            }
+
+            var result = await _userManager.DeleteAsync(user);
+            if (!result.Succeeded)
```
Is it still correct for a user to be signed in if this fails?

HaoK replied:
The default implementation basically never will fail,  it would only blow up with an exception, but this is a valid point, I guess we should only sign out if it succeeded

pranavkm commented on src/UI/Areas/Identity/Pages/Account/Manage/DeletePersonalData.cshtml.cs:
```diff
@@ -0,0 +1,61 @@
+            var result = await _userManager.DeleteAsync(user);
+            if (!result.Succeeded)
+            {
+                throw new ApplicationException($"Unexpected error occurred deleteing user with ID '{user.Id}'.");
```
`ApplicationException` -> `InvalidOperationException`

pranavkm commented on src/UI/Areas/Identity/Pages/Account/Manage/Privacy.cshtml:
```diff
@@ -0,0 +1,23 @@
+@model PrivacyModel
+@{
+    ViewData["Title"] = "Privacy";
+    ViewData["ActivePage"] = "Privacy";
```
`ManageNavPages.Privacy`

pranavkm commented on src/UI/Areas/Identity/Pages/Account/Manage/DownloadPersonalData.cshtml.cs:
```diff
@@ -0,0 +1,52 @@
+            _logger = logger;
+        }
+
+        public async Task<IActionResult> OnGet()
```
When do we get to this handler? `Privacy.cshtml` posts to this action and there isn't a link to this page.

HaoK replied:
Yeah we don't, I'll remove

mkArtakMSFT commented on src/UI/Areas/Identity/Pages/Account/Manage/DeletePersonalData.cshtml.cs:
```diff
@@ -0,0 +1,61 @@
+
+            await _signInManager.SignOutAsync();
+
+            _logger.LogInformation("User with ID '{UserId}' deleted themselves.", _userManager.GetUserId(User));
```
Is it ok to log the userId as is (as this can be the email address, which is PII)

HaoK replied:
The UserName is the email, the user id is a guid by default, but this is the pattern the existing templates use for logging I believe.

mkArtakMSFT replied:
Thanks!

mkArtakMSFT commented on src/UI/Areas/Identity/Pages/Account/Manage/Privacy.cshtml:
```diff
@@ -0,0 +1,23 @@
+</div>
+
+@section Scripts {
+    @await Html.PartialAsync("_ValidationScriptsPartial")
```
Similar to the rest of the comments - prefer <partial ... /> tag syntax instead

HaoK commented on src/UI/Areas/Identity/Pages/Account/Manage/DeletePersonalData.cshtml:
```diff
@@ -0,0 +1,28 @@
+        <strong>This action will delete your user.</strong>
+    </p>
+    <p>
+        This is unrecoverable.
```
"This action will delete your user."
This is unrecoverable.

means you can't recover your user after you delete it.

pranavkm replied:
Should this elaborate and say, once deleted you cannot use your login credentials and would need to re-register?

Kinda unrelated, but is there a hook that let's the app developer also clean up other stored details about the user when this is deleted? For instance, if the user id is used to store my order history or posted comments?

HaoK replied:
Don't worry about the exact text yet, I'll update that shortly.

re hook, not really, ideally we'd have an easy way for them to just drop in a new DeletePersonalData.cshtml.cs to take over this work entirely to customize.  Or I guess scaffold it all and change the code if that feature doesn't make it...

HaoK commented:
Ok updated switching all default UI over to use the new partial and consistently use ManageNavClass constants in for the active page stuff

HaoK commented:
Text has been updated, Privacy => Personal Data, will add one more change which will require users to reenter their password if they have a local password before being allowed to delete their account

HaoK commented:
Updated flow with password confirmation on Delete page if the user has a password.  Also added white list to only include Id/UserName/Email/Phone/and Confirmed properties)

HaoK commented:
I think all feedback has been addressed now, added `Response.Headers.Add("Content-Disposition", "attachment; filename=PersonalData.json");` to force download.

@pranavkm can you do a final review/sign off?

pranavkm commented on src/UI/Areas/Identity/Pages/Account/Manage/DeletePersonalData.cshtml.cs:
```diff
@@ -0,0 +1,84 @@
+            {
+                if (!await _userManager.CheckPasswordAsync(user, Input.Password))
+                {
+                    ModelState.AddModelError(string.Empty, "Password not correct.");
```
`AddModelError("Input.Password", `) if we'd like to show error along the password field.

HaoK commented:
f1ed482c26df6e242b1e4eca0948dc77ba5b8fa9

grahamehorner commented:
If a user has two factor authentication enable would the second factor used for verification before deletion?

HaoK commented:
No, we just check password right now, checking a second factor is significantly more involved, but its certainly doable if an app wants that, you basically just need to duplicate the two factor login flow
